### PR TITLE
[Batch Mode] Pass PrimarySpecificPaths through compiler. (4)

### PR DESCRIPTION
**Pull request by davidungar**
Batch mode requires that supplementary output paths be specific to the primary input being compiled.
This PR adds arguments and replaces existing output file arguments with an instance of PrimarySpecificPaths in order to make all outputs, not just the main output, dependent on the primary.
It goes along with an lldb PR. Must be merged with https://github.com/apple/swift-lldb/pull/290

Two pieces of state have been removed from IRGenOptions, since they might vary from primary to primary. Because parallel IR generation still needs all of the main output filenames, a parameter has been added, `parallelOutputFilenames`. I don't have a better solution, but it is a bit hokey since it has to be passed down several layers and only used for the parallel case.

### CODE REVIEW

**Comment by davidungar** [comment]
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
[Batch Mode] WIP: Pass PrimarySpecificPaths to createEmptyModule and IRGenModule() #290

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test linux platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Rebased to result of landing prior PR.

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
https://github.com/apple/swift/pull/14230

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/290

@swift-ci Please smoke test os x platform

File: include/swift/Basic/PrimarySpecificPaths.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
+//===--- PrimarySpecificPaths -----------------------------------*- C++ -*-===//
```
Comment (jrose-apple) [inline include/swift/Basic/PrimarySpecificPaths.h]: Nitpick: missing .h.

(Once upon a time these headers were for printouts. I know, right?)

File: include/swift/Frontend/Frontend.h [hunk 2]
```diff
@@ -301,6 +301,9 @@ class CompilerInvocation {
   bool hasSerializedAST() {
     return FrontendOpts.InputKind == InputFileKind::IFK_Swift_Library;
   }
+
+  PrimarySpecificPaths getPSPsForAtMostOnePrimary() const;
```
Comment (jrose-apple) [inline include/swift/Frontend/Frontend.h]: No acronyms in function names, please.

File: include/swift/Frontend/Frontend.h [hunk 3]
```diff
@@ -578,6 +581,11 @@ class CompilerInstance {
                                  OptionSet<TypeCheckingFlags> TypeCheckOptions);
 
   void finishTypeChecking(OptionSet<TypeCheckingFlags> TypeCheckOptions);
+
+public:
+  PrimarySpecificPaths getPSPsForWMO() const;
```
Comment (jrose-apple) [inline include/swift/Frontend/Frontend.h]: WMO *might* be an exception to the acronyms rule because most of the team already knows it, but even that is iffy, especially without a comment.

File: include/swift/SIL/SILModule.h [hunk 4]
```diff
@@ -246,7 +247,7 @@ class SILModule {
   // Intentionally marked private so that we need to use 'constructSIL()'
   // to construct a SILModule.
   SILModule(ModuleDecl *M, SILOptions &Options, const DeclContext *associatedDC,
-            bool wholeModule);
+            bool wholeModule, PrimarySpecificPaths PSPs);
```
Comment (jrose-apple) [inline include/swift/SIL/SILModule.h]: This is definitely not the way we want to go. A SILModule is intended to be independent of inputs, just like the AST representation of a source file is only very loosely connected to the buffer and underlying file-on-disk. It's fine for SIL*Gen* to have knowledge of how the module is getting created, but this seems like a violation of purpose.

File: lib/Frontend/FrontendInputsAndOutputs.cpp [hunk 5]
```diff
@@ -379,3 +380,17 @@ bool FrontendInputsAndOutputs::hasDependencyTrackerPath() const {
   return hasDependenciesPath() || hasReferenceDependenciesPath() ||
          hasLoadedModuleTracePath();
 }
+
+PrimarySpecificPaths
+FrontendInputsAndOutputs::getPSPsForAtMostOnePrimary() const {
```
Comment (jrose-apple) [inline lib/Frontend/FrontendInputsAndOutputs.cpp]: I don't know if it really makes sense to have a *mental* model of WMO as "zero primaries". That is, this strikes me as a weird function name. What are you actually trying to do?

File: lib/FrontendTool/FrontendTool.cpp [hunk 6]
```diff
@@ -291,8 +291,9 @@ static bool emitSyntax(SourceFile *SF, LangOptions &LangOpts,
 
 /// Writes SIL out to the given file.
 static bool writeSIL(SILModule &SM, ModuleDecl *M, bool EmitVerboseSIL,
-                     StringRef OutputFilename, bool SortSIL) {
-  auto OS = getFileOutputStream(OutputFilename, M->getASTContext());
+                     bool SortSIL) {
+  auto OS =
+      getFileOutputStream(SM.getPSPs().OutputFilename, M->getASTContext());
```
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: Yeah, in theory you ought to be able to write the same SIL module multiple places. I can't see why someone would ever do that, but it shouldn't be baked in.

**Review by jrose-apple — COMMENTED** [review]
Didn't review all the little bits because I ended up with a big objection: paths do not belong in the SILModule. A bit more elaboration in the specific comment for that.

File: include/swift/SIL/SILModule.h (hunk 4 above)
Comment (davidungar) [reply include/swift/SIL/SILModule.h]: Would you like me to investigate an alternative in which the paths are not part of the SILModule, but rather passed along as parameters where they are needed instead?

**Comment by davidungar** [comment]
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/290

@swift-ci Please smoke test os x platform

File: include/swift/Basic/PrimarySpecificPaths.h (hunk 1 above)
Comment (davidungar) [reply include/swift/Basic/PrimarySpecificPaths.h]: Fixed.

File: include/swift/Frontend/Frontend.h (hunk 2 above)
Comment (davidungar) [reply include/swift/Frontend/Frontend.h]: Fixed.

File: include/swift/Frontend/Frontend.h (hunk 3 above)
Comment (davidungar) [reply include/swift/Frontend/Frontend.h]: Spelled it out.

File: include/swift/SIL/SILModule.h (hunk 4 above)
Comment (davidungar) [reply include/swift/SIL/SILModule.h]: Have moved the paths out of SILModule and into parameters.

File: lib/Frontend/FrontendInputsAndOutputs.cpp (hunk 5 above)
Comment (davidungar) [reply lib/Frontend/FrontendInputsAndOutputs.cpp]: There are cases where we won't support batch mode with > 1 primary. For example, when compiling a SILModule instead of a source file. In such cases, we can get the PSPs without knowing the InputFile for which code is being generated. That's what this function is doing: getting the primary-specific paths when there are not a multiplicity of primaries.

Here is the implementation:

    return PrimarySpecificPaths(        
                              getSingleOutputFilename(),
                              hasInputsProducingMainOutputs()
                                  ? firstInputProducingOutput().file()
                                  : StringRef(),
                              supplementaryOutputs());

Basically it gets the same information that would have been used before batch mode.

**Comment by davidungar** [comment]
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please clean smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please clean smoke test linux platform

**Comment by dcci** [comment]
@davidungar thanks! This is exactly the UB I found, i.e. `DependencyTracker` might `nullptr` when dereferenced. I think this also shows up on `master` (i.e. without your PR), so maybe it makes sense/you might consider to split the last patch and push it separately?

**Comment by davidungar** [comment]
@dcci Great point! I will look over master and submit a PR to handle all the times the dependency tracker is a null pointer. Will have to decide whether to fail an assert, emit a diagnostic, or what.

**Comment by davidungar** [comment]
@dcci After inspecting the code, I believe that the null pointer won't ever be dereferenced, so I have submitted https://github.com/apple/swift/pull/14543 to fix the (static) issue in the master branch.

**Comment by davidungar** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please clean smoke test

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test

File: include/swift/IRGen/IRGenPublic.h [hunk 7]
```diff
@@ -26,7 +26,8 @@ class IRGenModule;
 
 /// Create an IRGen module.
 std::pair<IRGenerator *, IRGenModule *>
-createIRGenModule(SILModule *SILMod, llvm::LLVMContext &LLVMContext);
+createIRGenModule(SILModule *SILMod, PrimarySpecificPaths PSPs,
```
Comment (jrose-apple) [inline include/swift/IRGen/IRGenPublic.h]: Seems like this should be passed by const reference (here and elsewhere in IRGen/Subsystems). It's not cheap to copy.

File: include/swift/Subsystems.h [hunk 8]
```diff
@@ -279,12 +281,13 @@ namespace swift {
 
   /// Wrap a serialized module inside a swift AST section in an object file.
   void createSwiftModuleObjectFile(SILModule &SILMod, StringRef Buffer,
-                                   StringRef OutputPath);
+                                   StringRef ModuleName,
+                                   const PrimarySpecificPaths PSPs);
```
Comment (jrose-apple) [inline include/swift/Subsystems.h]: Stray `const` (or was it meant to be a reference?)

File: lib/Frontend/FrontendInputsAndOutputs.cpp [hunk 9]
```diff
@@ -303,6 +308,18 @@ void FrontendInputsAndOutputs::setMainAndSupplementaryOutputs(
       AllInputs[i].setOutputFilename(outputFiles[i]);
   }
   SupplementaryOutputs = supplementaryOutputs;
+
+  if (hasInputs() && primaryInputCount() < 2) {
```
Comment (jrose-apple) [inline lib/Frontend/FrontendInputsAndOutputs.cpp]: I feel like it's generally weird to coalesce "whole-module" and "only one primary" like this. I'd prefer if you just ask for both of those conditions separately.

File: lib/FrontendTool/FrontendTool.cpp [hunk 10]
```diff
@@ -1247,24 +1254,25 @@ static bool performCompileStepsPostSILGen(CompilerInstance &Instance,
...
+      !PSPs.SupplementaryOutputs.ModuleDocOutputPath.empty();
   if (haveModulePath && !SM->isSerialized())
     SM->serialize();
 
   if (haveModulePath) {
     if (Action == FrontendOptions::ActionType::MergeModules ||
         Action == FrontendOptions::ActionType::EmitModuleOnly) {
-      return emitIndexDataIfNeeded(Instance.getPrimarySourceFile(), Invocation,
+      // FIXME: what if MSF is a module?
+      return emitIndexDataIfNeeded(MSF.dyn_cast<SourceFile *>(), Invocation,
```
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: `emitIndexDataIfNeeded` already handles that case; it'll index everything. You could change it to take a ModuleOrSourceFile if it's bothering you. (It's currently getting the module out of the CompilerInstance, which isn't really the best.)

File: lib/FrontendTool/FrontendTool.cpp [hunk 11]
```diff
@@ -1294,12 +1302,12 @@ static bool performCompileStepsPostSILGen(CompilerInstance &Instance,
     return processCommandLineAndRunImmediately(
         Invocation, Instance, std::move(SM), MSF, observer, ReturnValue);
 
+  const std::string OutputFilename = PSPs.OutputFilename;
```
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: No need to eagerly copy this, right? (StringRef)

File: lib/SILGen/SILGen.cpp [hunk 12]
```diff
@@ -1485,8 +1483,7 @@ SILModule::constructSIL(ModuleDecl *mod, SILOptions &options, FileUnit *SF,
 }
 
 std::unique_ptr<SILModule>
-swift::performSILGeneration(ModuleDecl *mod,
-                            SILOptions &options,
+swift::performSILGeneration(ModuleDecl *mod, SILOptions &options,
```
Comment (jrose-apple) [inline lib/SILGen/SILGen.cpp]: In the future, please try to avoid changes that are just reformatting in a file you aren't otherwise editing.

**Review by jrose-apple — COMMENTED** [review]
Looking pretty good!

File: lib/SILGen/SILGen.cpp (hunk 12 above)
Comment (davidungar) [reply lib/SILGen/SILGen.cpp]: So I checked, and these changes were made by git-clang-format. I understand that they make it harder to review code, but also understand that the uniformity obtained by git-clang-format can make it easier. I had thought that the best practice was to use git-clang-format, but from your comment I'm surmising that best practice would be to use git-clang-format, then undo any changes that it made and still stay within 80-characters. Is that correct?

Concerning this particular file, I had changed it and then changed it back. Should have then undone the git-clang-formatting changes. But let me know which practice would serve the team best in the future. Thanks!

For this PR, I will go through all my changes and undo all similar ones.

File: lib/Frontend/FrontendInputsAndOutputs.cpp (hunk 9 above)
Comment (davidungar) [reply lib/Frontend/FrontendInputsAndOutputs.cpp]: Changed to `hasUniquePrimaryInput() || (hasInputs() && isWholeModule())`

File: lib/SILGen/SILGen.cpp (hunk 12 above)
Comment (jrose-apple) [reply lib/SILGen/SILGen.cpp]: The best practice is to only format the areas of code that you touch, not the whole file, but having the rest of the file get cleaned up is all right as well. But yes, that last part is the important part: if you end up *not* changing a file, please remove any formatting changes as well; we can do them some other time.

File: lib/FrontendTool/FrontendTool.cpp (hunk 11 above)
Comment (davidungar) [reply lib/FrontendTool/FrontendTool.cpp]: If, at some time in the future, Invocation is discarded before code generation is finished, using a StringRef would introduce a bug. Do you want to depend on the Invocation not being freed?

File: lib/FrontendTool/FrontendTool.cpp (hunk 10 above)
Comment (davidungar) [reply lib/FrontendTool/FrontendTool.cpp]: Changed the FIXME to a comment including your answer.

File: include/swift/IRGen/IRGenPublic.h (hunk 7 above)
Comment (davidungar) [reply include/swift/IRGen/IRGenPublic.h]: Copying avoids reliance on FrontendInputsAndOutputs not getting freed. Do you think that saving cost of copying perhaps a dozen strings is worth an implicit lifetime dependency?

Comment (davidungar) [reply include/swift/IRGen/IRGenPublic.h]: Or is the dependency already explicit?

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test

File: lib/Frontend/FrontendInputsAndOutputs.cpp (hunk 9 above)
Comment (davidungar) [reply lib/Frontend/FrontendInputsAndOutputs.cpp]: Tx

File: include/swift/Subsystems.h (hunk 8 above)
Comment (davidungar) [reply include/swift/Subsystems.h]: This parameter is now a StringRef. Sorry for the interim confusion.

File: lib/SILGen/SILGen.cpp (hunk 12 above)
Comment (davidungar) [reply lib/SILGen/SILGen.cpp]: OK, thanks a lot for clarifying! I may not be running git-clang-format in the optimum fashion. I think I've now caught all of those.

File: lib/FrontendTool/FrontendTool.cpp (hunk 11 above)
Comment (jrose-apple) [reply lib/FrontendTool/FrontendTool.cpp]: I don't think that could happen, since `generateIR` doesn't get a reference to the Invocation. (Also, PSPs gets passed to `generateIR` right now anyway.)

File: include/swift/IRGen/IRGenPublic.h (hunk 7 above)
Comment (jrose-apple) [reply include/swift/IRGen/IRGenPublic.h]: It's sort of explicit? SILModule -> ASTContext -> DiagnosticEngine -> CompilerInstance. But mostly we just assume it's better *not* to copy something when passing information down the stack—almost all lifetimes are stack-oriented. If you do a copy anywhere, it would be at the point where you move the PSPs into the IRGenModule, because that gets persisted beyond the constructor call.

**Review by graydon — COMMENTED** [review]
Aside from Jordan's point that maybe PSPs lifetime makes it viable to pass down-stack as a const&, this looks like uncontroversial refactoring to me!

File: include/swift/IRGen/IRGenPublic.h (hunk 7 above)
Comment (davidungar) [reply include/swift/IRGen/IRGenPublic.h]: OK, will fix createIRGenModule to take and pass down a const ref for PSPs.

Comment (davidungar) [reply include/swift/IRGen/IRGenPublic.h]: Thanks for explaining.

File: lib/FrontendTool/FrontendTool.cpp (hunk 11 above)
Comment (davidungar) [reply lib/FrontendTool/FrontendTool.cpp]: Thanks for explaining. I've fixed this and also parallelOutputFilenames to avoid copying as they go down the stack.

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test

File: include/swift/IRGen/IRGenPublic.h (hunk 7 above)
Comment (davidungar) [reply include/swift/IRGen/IRGenPublic.h]: I have made some changes to pass down const refs.

**Comment by davidungar** [comment]
Thank you, @graydon.

File: lib/IRGen/IRGen.cpp [hunk 13]
```diff
@@ -893,7 +897,7 @@ static void performParallelIRGeneration(IRGenOptions &Opts,
     
     // There must be an output filename for each source file.
     // We ignore additional output filenames.
-    if (OutputIter == Opts.OutputFilenames.end()) {
+    if (OutputIter == outputFilenames.end()) {
       // TODO: Check this already at argument parsing.
```
Comment (jrose-apple) [inline lib/IRGen/IRGen.cpp]: Hey, I think you fixed this TODO by now!

File: lib/IRGen/IRGen.cpp [hunk 14]
```diff
@@ -909,7 +913,10 @@ static void performParallelIRGeneration(IRGenOptions &Opts,
     // Create the IR emitter.
     IRGenModule *IGM = new IRGenModule(irgen, std::move(targetMachine),
                                        nextSF, *Context,
-                                       ModuleName, *OutputIter++);
+                                       ModuleName,
+                                       PrimarySpecificPaths(*OutputIter++,
+                                                          nextSF->getFilename(),
+                                       PSPs.SupplementaryOutputs));
```
Comment (jrose-apple) [inline lib/IRGen/IRGen.cpp]: At this point we don't really support any supplementary outputs, so you might as well make an empty set, right?

Comment (jrose-apple) [reply lib/IRGen/IRGen.cpp]: (because otherwise you'd just stomp on the same files over and over)

File: lib/IRGen/IRGen.cpp [hunk 15]
```diff
@@ -1069,32 +1076,36 @@ static void performParallelIRGeneration(IRGenOptions &Opts,
...
 performIRGeneration(IRGenOptions &Opts, swift::ModuleDecl *M,
                     std::unique_ptr<SILModule> SILMod,
-                    StringRef ModuleName, llvm::LLVMContext &LLVMContext,
+                    StringRef ModuleName, const PrimarySpecificPaths &PSPs,
+                    llvm::LLVMContext &LLVMContext,
+                    ArrayRef<std::string> parallelOutputFilenames,
                     llvm::GlobalVariable **outModuleHash) {
-  if (SILMod->getOptions().shouldPerformIRGenerationInParallel()) {
+  if (SILMod->getOptions().shouldPerformIRGenerationInParallel() &&
+      !parallelOutputFilenames.empty()) {
```
Comment (jrose-apple) [inline lib/IRGen/IRGen.cpp]: Worth asserting instead of testing?

File: lib/SIL/SILModule.cpp [hunk 16]
```diff
@@ -15,6 +15,7 @@
 #include "swift/AST/GenericEnvironment.h"
 #include "swift/AST/ProtocolConformance.h"
 #include "swift/AST/Substitution.h"
+#include "swift/Basic/PrimarySpecificPaths.h"
```
Comment (jrose-apple) [inline lib/SIL/SILModule.cpp]: Doesn't seem to be used anymore.

**Review by jrose-apple — COMMENTED** [review]
Only scattered comments by this point.

File: lib/SIL/SILModule.cpp (hunk 16 above)
Comment (davidungar) [reply lib/SIL/SILModule.cpp]: Thanks! Removed.

File: lib/IRGen/IRGen.cpp (hunk 15 above)
Comment (davidungar) [reply lib/IRGen/IRGen.cpp]: I'll try it, but the callers that invoke this for REPL (for instance), may fail if NumThreads is set.

Comment (davidungar) [reply lib/IRGen/IRGen.cpp]: It may be possible to fail the assertion if the command line sets -num-threads but calls for immediate execution. I could modify the callers on that path to pass in the right file names. What do you think?

File: lib/IRGen/IRGen.cpp (hunk 14 above)
Comment (davidungar) [reply lib/IRGen/IRGen.cpp]: What if -num-threads 1? Useful in that case??

File: lib/IRGen/IRGen.cpp (hunk 13 above)
Comment (davidungar) [reply lib/IRGen/IRGen.cpp]: Cool! Removed.

File: lib/FrontendTool/FrontendTool.cpp [hunk 17]
```diff
@@ -1053,9 +1066,10 @@ static void setPrivateDiscriminatorIfNeeded(IRGenOptions &IRGenOpts,
 }
 
 static bool serializeSIB(FrontendOptions &opts, SILModule *SM,
-                         ASTContext &Context, ModuleOrSourceFile MSF) {
+                         const PrimarySpecificPaths &PSPs, ASTContext &Context,
+                         ModuleOrSourceFile MSF) {
   const std::string &moduleOutputPath =
-      opts.InputsAndOutputs.supplementaryOutputs().ModuleOutputPath;
+      PSPs.SupplementaryOutputs.ModuleOutputPath;
```
[the remark this answers is not in the document]
Comment (davidungar) [reply lib/FrontendTool/FrontendTool.cpp]: Thanks for catching this. Fixed.

File: include/swift/Frontend/FrontendInputsAndOutputs.h [hunk 18]
```diff
@@ -210,6 +221,17 @@ class FrontendInputsAndOutputs {
   void forEachInputProducingSupplementaryOutput(
       llvm::function_ref<void(const InputFile &)> fn) const;
 
+  /// Assumes there is not more than one primary input file, if any.
+  /// Otherwise, you would need to call getPrimarySpecificPathsForPrimary
+  /// to tell it which primary input you wanted the outputs for.
+  ///
+  /// Must not be constructed on-the-fly because some parts of the compiler
+  /// truck in StringRef's to its components, so it must live as long as the
```
[the remark this answers is not in the document]
Comment (davidungar) [reply include/swift/Frontend/FrontendInputsAndOutputs.h]: I meant it in the sense of the second definition 11 in www.dictionary.com/browse/truck?s=t : "to have dealings" or second definition 10: "to exchange commodities, barter". But, since it is unclear--maybe that usage has gone out of fashion?--I'll 86 it. The rewording is clearer.

Regarding the apostrophe, it was there to show the plural of an abbreviation, as mentioned in www.grammar-monster.com/lessons/apostrophes_show_plural_of_abbreviations.htm .
But I'll take it out, since I assume it is not common usage in this code base.

File: include/swift/Frontend/FrontendInputsAndOutputs.h [hunk 19]
```diff
@@ -210,6 +221,17 @@ class FrontendInputsAndOutputs {
...
+  /// compiler.
+  PrimarySpecificPaths &getPrimarySpecificPathsForAtMostOnePrimary();
+
+  PrimarySpecificPaths &getPrimarySpecificPathsForPrimary(StringRef);
```
[the remark this answers is not in the document]
Comment (davidungar) [reply include/swift/Frontend/FrontendInputsAndOutputs.h]: Fixed. Thanks. It *is* clearer with the name.

**Comment by davidungar** [comment]
Thanks for getting to these today! As you can tell from the responses, I'm too tired to do it justice tonight, but have started in and expect to test and push tomorrow. As you can see, I have a couple of questions about the parallelOutputFiles and supplementaryPaths arguments in the parallel case. Let me know your thoughts if you like and/or I'll sleep on it and think about it tomorrow. These are interesting issues.

File: lib/IRGen/IRGen.cpp (hunk 15 above)
Comment (jrose-apple) [reply lib/IRGen/IRGen.cpp]: Ah, interesting. All right, I guess it makes sense as is.

File: lib/IRGen/IRGen.cpp (hunk 14 above)
Comment (jrose-apple) [reply lib/IRGen/IRGen.cpp]: I don't think it's important enough to support when we have a perfectly good "don't parallelize" mode. If we ever unify those then it might be more interesting.

File: include/swift/Frontend/FrontendInputsAndOutputs.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (jrose-apple) [reply include/swift/Frontend/FrontendInputsAndOutputs.h]: Ah, I guess I've heard "traffic in" for that idea more.

And right, while the abbreviation apostrophe is controversial, I think the deciding factor here is that we're not considering complete type names to be abbreviations, even if they have abbreviations in them.

File: lib/IRGen/IRGen.cpp (hunk 15 above)
Comment (davidungar) [reply lib/IRGen/IRGen.cpp]: OK, thanks. Will keep it as is.

File: lib/IRGen/IRGen.cpp (hunk 14 above)
Comment (davidungar) [reply lib/IRGen/IRGen.cpp]: OK, I will pass in empty supplementary outputs here (by default).
Also the PSPs parameter to the enclosing function becomes unused and so I'm deleting it.
*But*, at the point where performParallelIRGeneration is called, I will add an assertion that the supplementary output paths are empty.

File: include/swift/Frontend/FrontendInputsAndOutputs.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (davidungar) [reply include/swift/Frontend/FrontendInputsAndOutputs.h]: Ah, thanks for the clarification about type names not considered as abbreviations. Yeah, I think "truck in" might be a usage from the 60s. Anyway, I reworded it.

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test

**Comment by davidungar** [comment]
I don't have time to do it now, but I'm wondering if parallelIRGeneration shouldn't support supplementary outputs. I suspect it needs to for one set.

**Comment by jrose-apple** [comment]
We don't currently have any supplementary outputs that IRGen is responsible for, so it probably just hasn't come up.

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test

**Comment by davidungar** [comment]
After more thought, I decided it was appropriate to use the supplemental outputs for only the first parallel-genererated input.

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test

**Comment by davidungar** [comment]
@jrose-apple At this point, I *think* all issues have been addressed. Will see if it passes the tests.

**Comment by jrose-apple** [comment]
> After more thought, I decided it was appropriate to use the supplemental outputs for only the first parallel-genererated input.

I'm not sure I like this. Which input is first is essentially arbitrary, and whatever output we would get from such a thing wouldn't have all the data. I'd really rather not pretend we can handle this when we can't.

@eeckstein, thoughts? We're talking about if IRGen were to have outputs in addition to its main output, like if it could output a .ll file but still go on to output a .o. In the case of parallel WMO, we've set everything *else* up so that there's only one of each supplemental output for the whole process, but that breaks down once we hit IRGen.

**Comment by davidungar** [comment]
Fair point. I'll look at NOT passing a PrimarySpecificPaths into the IRGenModule constructor. If it doesn't emit any supplementaries, then it should only receive an output filename, and a main input filename (for debug info).

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test

**Comment by davidungar** [comment]
@jrose-apple OK, this version should be cleaner. (Thanks for your patience.) Since IRGenModule does nothing with the supplementaries, I think it may make sense to not pass in supplementary output names at all to its constructor. I've also tried to reformat in the latest version to minimize the diffs.
If it looks good and you would like me to streamline the commits, I can squash some. Thanks.

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please clean smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please clean smoke test os x platform

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please clean smoke test os x platform

File: tools/sil-opt/SILOpt.cpp [hunk 20]
```diff
@@ -426,8 +426,10 @@ int main(int argc, char **argv) {
     runSILLoweringPasses(*CI.getSILModule());
   } else {
     auto *SILMod = CI.getSILModule();
-    {
-      auto T = irgen::createIRGenModule(SILMod, getGlobalLLVMContext());
+    { const auto PSPs = CI.getPrimarySpecificPathsForAtMostOnePrimary();
```
Comment (jrose-apple) [inline tools/sil-opt/SILOpt.cpp]: Missing newline here!

**Review by jrose-apple — APPROVED** [review]
I think this is it!

File: tools/sil-opt/SILOpt.cpp (hunk 20 above)
Comment (davidungar) [reply tools/sil-opt/SILOpt.cpp]: Fixed! Great to have worked together and figured out how to do this right. Thanks!

**Comment by davidungar** [comment]
Please test with the following PR:
apple/swift-lldb#290

@swift-ci Please clean smoke test
